### PR TITLE
Introduce --stablehlo-canonicalize-dynamism

### PR DESCRIPTION
**Pull request by burmako**
This pull request achieves two goals:

1) Makes specialization of dynamic StableHLO programs completely independent from MHLO. Before this PR, we were depending on StableHLO => MHLO and then MHLO canonicalization to actually specialize away dynamic_foo ops, but not anymore.

2) Provides a platform for addressing #1367. As a follow-up to this PR, we will add another pattern to this pass that transforms CustomCallOp to erase the indices_of_shape_operands attributes as well as the operands called out in indices_of_shape_operands.

The patterns in this PR are quite straightforward, since almost all of the work is done in --stablehlo-refine-shapes.
CanonicalizeRealDynamicSliceOpToDynamicSliceOpPattern is the only exception because it isn't just about "the result has static shape, let's go" or "the attributes are all static, let's go".

Furthermore, this PR extends some existing verifiers and adds some new verifiers for dynamic ops. Since dynamic ops aren't specced yet, it is expected that they have implementation gaps, but in this case, these gaps would've allowed invalid dynamic ops to get transformed into valid static ops, so I fixed them right away.

### CODE REVIEW

File: stablehlo/dialect/TypeInference.cpp [hunk 1]
```diff
@@ -3471,13 +3499,22 @@ LogicalResult verifyDynamicPadOp(std::optional<Location> location,
 
 LogicalResult verifyDynamicReshapeOp(std::optional<Location> location,
                                      Value outputShape, Value result) {
-  auto resultType = result.getType().dyn_cast<RankedTensorType>();
-  auto outputShapeType = outputShape.getType().dyn_cast<RankedTensorType>();
-  if (resultType && outputShapeType && outputShapeType.hasStaticShape() &&
+  auto resultType = result.getType().dyn_cast<ShapedType>();
+  auto outputShapeType = outputShape.getType().dyn_cast<ShapedType>();
+  if (resultType.hasRank() && outputShapeType.hasStaticShape() &&
```
Comment (GleasonK) [inline stablehlo/dialect/TypeInference.cpp:3504]: Deleted the `outputShapeType` and `resultType` null check? Should `dyn_cast` be `cast` if this is guaranteed?

File: stablehlo/dialect/TypeInference.cpp [hunk 2]
```diff
@@ -3471,13 +3499,22 @@ LogicalResult verifyDynamicPadOp(std::optional<Location> location,
...
                              "elements in output_shape");
+  SmallVector<int64_t> outputShapeVec;
+  if (succeeded(matchInts(outputShape, outputShapeVec))) {
+    auto inferredResultType =
+        RankedTensorType::get(outputShapeVec, resultType.getElementType());
+    if (!isCompatibleForHloTypeInference(inferredResultType, resultType))
+      return emitOptionalError(
+          location, "inferred type ", inferredResultType,
+          " is incompatible with return type of operation ", resultType);
+  }
```
Comment (GleasonK) [inline stablehlo/dialect/TypeInference.cpp:3517]: This block of code appears 3 times in this PR, is it worth making this generic?

**Review by GleasonK — APPROVED** [review]

File: stablehlo/dialect/TypeInference.cpp (hunk 1 above)
Comment (burmako) [reply stablehlo/dialect/TypeInference.cpp:3504]: Yes, this is guaranteed. Good catch that this should be a `cast` - we actually have quite a few places in our codebase which are in a similar situation which I have all fixed now.

File: stablehlo/dialect/TypeInference.cpp (hunk 2 above)
Comment (burmako) [reply stablehlo/dialect/TypeInference.cpp]: Done.

**Pull request merged**